fix(hlsl-out): use `Interlocked<op>` intrinsic for atomic integers

ErichDonGubler opened the pull request:
Fixes #2284.

---

We currently assume that we are using raw `RWByteAddressBuffer` methods for all atomic operations (`<pointer>.Interlocked<op>(<raw_byte_offset>, …)`), which is only true when we use `var<storage, read_write>` globals. For `var<workgroup>` globals, we need `Interlocked<op>(<pointer>, …)`, using the original expression as the first argument.

Fix this by branching on the `pointer`'s address space in `Atomic` statements, and implementing the workgroup address space case with intrinsics.

ErichDonGubler commented on src/back/hlsl/writer.rs:
```diff
@@ -1856,14 +1856,66 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
                     }
                 };
 
+                let pointer_is_atomic_scalar = {
+                    fn is_atomic_scalar_ty<'a>(
+                        module: &'a Module,
+                        func_ctx: &'a back::FunctionCtx,
+                        ty: Handle<crate::Type>,
+                    ) -> bool {
+                        match &module.types[ty].inner {
+                            TypeInner::Atomic {
+                                // TODO: floats and booleans aren't valid to interlock, so uh…problem?
+                                kind: crate::ScalarKind::Sint | crate::ScalarKind::Uint,
+                                ..
+                            } => true,
+                            TypeInner::Array { base, .. } | TypeInner::Pointer { base, .. } => {
+                                is_atomic_scalar_ty(module, func_ctx, *base)
+                            }
+                            _ => false,
+                        }
+                    }
+
+                    fn is_atomic_scalar_expr<'a>(
+                        module: &'a Module,
+                        func_ctx: &'a back::FunctionCtx,
+                        expr: Handle<crate::Expression>,
+                    ) -> bool {
+                        match &func_ctx.expressions[expr] {
+                            crate::Expression::GlobalVariable(handle) => is_atomic_scalar_ty(
+                                module,
+                                func_ctx,
+                                module.global_variables[*handle].ty,
+                            ),
+                            crate::Expression::Access { base, .. }
+                            | crate::Expression::AccessIndex { base, .. } => {
+                                is_atomic_scalar_expr(module, func_ctx, *base)
+                            }
+                            _ => false,
+                        }
+                    }
```
todo: What expressions and types have been missed that should be permitted?

ErichDonGubler commented on src/back/hlsl/writer.rs:
```diff
@@ -1856,14 +1856,66 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
+                    ) -> bool {
+                        match &module.types[ty].inner {
+                            TypeInner::Atomic {
+                                // TODO: floats and booleans aren't valid to interlock, so uh…problem?
+                                kind: crate::ScalarKind::Sint | crate::ScalarKind::Uint,
```
todo: We actually _should_ be handling floats and booleans here, I'm guessing, but support doesn't seem straightforward to me in my (very limited) search:

* Floats need a specific shader model version or above, not sure which yet. I've seen mention in shader model 6 docs, but that might not be the minimum version.
* No idea about boolean, but I haven't seen it mentioned in any docs I've seen so far.

jimblandy replied:
`ScalarKind` is indeed too loose for this particular operation. We forbid types other than i32 and u32 [here](https://github.com/gfx-rs/naga/blob/52043be00489d6dedb08c36bb97d018d11cb1e13/src/valid/type.rs#L306-L309).

ErichDonGubler replied:
Gotcha. Should be resolvable, then. I removed my `TODO` comment, and I think otherwise this code should be correct for the question I had.

ErichDonGubler commented:
@jimblandy and I are planning on pair-reviewing this draft Soon™. 😊

ErichDonGubler commented on tests/in/atomicSimple.wgsl:
```diff
@@ -0,0 +1,49 @@
+
+@compute
+@workgroup_size(2)
+fn corge(@builtin(local_invocation_id) id: vec3<u32>) {
```
nitpick: `corge` is random gibberish that served its purpose while I was still spending brain cycles understanding other things. Maybe something more sensible (even if bland, like `cs_main`) should be used here?

ErichDonGubler replied:
Renamed to `cs_main`.

ErichDonGubler commented on tests/out/glsl/atomicSimple.corge.Compute.glsl:
```diff
@@ -0,0 +1,57 @@
+
+
+void main() {
+    if (gl_LocalInvocationID == uvec3(0u)) {
+        foo = 0u;
+        baz = uint[2](0u, 0u);
+    }
```
question: I'm actually puzzled, but have a hypothesis as to why this code gets emitted, though there's no apparent equivalent in WGSL source. This appears to be automatically generated init., so variables have a well-known value. Does invocation `0` always run first? Is that right?

ErichDonGubler replied:
Bringing back in context from sync discussion with @jimblandy: This is actually a _tie-breaker_ for the workgroup, where we elect one and only one thread/invocation in the thread group/workgroup to set the group's state properly.

I'm satisfied with what I've learned here. Thanks, Jim!

ErichDonGubler replied:
Interestingly, Raph Levien [notes](https://stackoverflow.com/questions/39393560/glsl-memorybarriershared-usefulness#comment119242328_39395947) that in OpenGL 4.6 and GL ES 3.2, the `memoryBarrierShared` might not be necessary here. IDK what our deps are, so 🤷🏻‍♂️. Just interesting trivia, I guess?

ErichDonGubler commented on tests/out/hlsl/atomicSimple.hlsl:
```diff
@@ -0,0 +1,40 @@
+[numthreads(2, 1, 1)]
+void corge(uint3 id : SV_GroupThreadID, uint3 __local_invocation_id : SV_GroupThreadID)
+{
+    if (all(__local_invocation_id == uint3(0u, 0u, 0u))) {
```
question: What is `all(…)` here? 🤔

ErichDonGubler replied:
Responding to my own question, looks like this is the [`all(…)` intrinsic](https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-all), which can be used on scalar, vectors, or matrices to determine that all components are non-zero. That…seems unnecessary here, but doesn't hurt anything, I suppose! Definitely nothing that should block here.

teoxoy replied:
`all` is required here since the `==` operator returns a boolean vector and the if statement expects a boolean (see https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-operators#comparison-operators).

ErichDonGubler commented on tests/snapshots.rs:
```diff
@@ -505,6 +505,10 @@ fn convert_wgsl() {
                 | Targets::ANALYSIS,
         ),
+        (
+            "atomicSimple",
```
Idea: `atomicOps`

ErichDonGubler replied:
Using `atomicOps` for now. Leaving this open to bikeshedding, but I'm not gonna spend more time on it without prompting.

ErichDonGubler commented on tests/out/glsl/atomicSimple.corge.Compute.glsl:
```diff
@@ -0,0 +1,57 @@
+    barrier();
+    uint unnamed = foo;
+    uint unnamed_1 = baz[0];
+    memoryBarrierShared();
+    barrier();
```
In the same link I discovered [elsewhere](https://github.com/gfx-rs/naga/pull/2294/files#r1153751514), I found an answer. Yippee!

ErichDonGubler commented on tests/out/glsl/atomicSimple.corge.Compute.glsl:
```diff
@@ -0,0 +1,57 @@
+    memoryBarrierShared();
+    barrier();
+    uint unnamed = foo;
+    uint unnamed_1 = baz[0];
```
Ahhh, crap, just did a switch to `baz[1]`, and it looks like byte offsets are still being used. Now I get to figure out how to fix it…hopefully not by duplicating too much of `Writer::fill_access_chain`. 😓

ErichDonGubler commented:
@jimblandy: I've pushed some significant updates to this during my afternoon yesterday. Most of the logic is the same, just renamed or shuffled around a bit.

ErichDonGubler commented on src/back/hlsl/writer.rs:
```diff
@@ -1856,14 +1856,54 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
-                // working around the borrow checker in `self.write_expr`
-                let chain = mem::take(&mut self.temp_access_chain);
-                let var_name = &self.names[&NameKey::GlobalVariable(var_handle)];
+                let accessing_byte_addr_buf = {
+                    fn is_atomic_scalar_ty(module: &Module, ty: Handle<crate::Type>) -> bool {
+                        match &module.types[ty].inner {
+                            &TypeInner::Atomic {
+                                kind: crate::ScalarKind::Sint | crate::ScalarKind::Uint,
+                                ..
+                            } => true,
+                            &TypeInner::Array { base, .. } | &TypeInner::Pointer { base, .. } => {
+                                is_atomic_scalar_ty(module, base)
+                            }
+                            _ => false,
+                        }
+                    }
+
+                    fn is_atomic_scalar_expr(
+                        module: &Module,
+                        func_ctx: &back::FunctionCtx,
+                        expr: Handle<crate::Expression>,
+                    ) -> bool {
+                        match &func_ctx.expressions[expr] {
+                            &crate::Expression::GlobalVariable(handle) => {
+                                is_atomic_scalar_ty(module, module.global_variables[handle].ty)
+                            }
+                            &crate::Expression::Access { base, .. }
+                            | &crate::Expression::AccessIndex { base, .. } => {
+                                is_atomic_scalar_expr(module, func_ctx, base)
+                            }
+                            _ => false,
+                        }
+                    }
```
question: Should the helper functions inside the block returning accessing_byte_addr_buf's value be moved to somewhere else?

ErichDonGubler replied:
This has been invalidated by recent changes. Dropping.

ErichDonGubler commented on src/back/hlsl/writer.rs:
```diff
@@ -1856,14 +1856,54 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
+                            &TypeInner::Array { base, .. } | &TypeInner::Pointer { base, .. } => {
+                                is_atomic_scalar_ty(module, base)
+                            }
+                            _ => false,
+                        }
+                    }
+
+                    fn is_atomic_scalar_expr(
+                        module: &Module,
+                        func_ctx: &back::FunctionCtx,
+                        expr: Handle<crate::Expression>,
+                    ) -> bool {
+                        match &func_ctx.expressions[expr] {
+                            &crate::Expression::GlobalVariable(handle) => {
+                                is_atomic_scalar_ty(module, module.global_variables[handle].ty)
+                            }
+                            &crate::Expression::Access { base, .. }
+                            | &crate::Expression::AccessIndex { base, .. } => {
+                                is_atomic_scalar_expr(module, func_ctx, base)
+                            }
+                            _ => false,
```
question: RE: the wildcard pattern lines a la `_ => false`: Are there any expression or type constructs that we need in this PR before merging?

ErichDonGubler commented:
@jimblandy: Just pushed up newest changes branching on address space. TIL a bunch, thanks for the help! ❤️

teoxoy commented on src/back/hlsl/writer.rs:
```diff
@@ -1856,14 +1856,44 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
-                // working around the borrow checker in `self.write_expr`
-                let chain = mem::take(&mut self.temp_access_chain);
-                let var_name = &self.names[&NameKey::GlobalVariable(var_handle)];
+                let pointer_space = {
+                    let pointer_ty = match func_ctx.info[pointer].ty {
+                        proc::TypeResolution::Handle(handle) => &module.types[handle].inner,
+                        proc::TypeResolution::Value(ref inner) => inner,
+                    };
+                    match pointer_ty {
+                        &TypeInner::Pointer { space, .. } => space,
+                        &TypeInner::ValuePointer { space, .. } => space,
+                        ref other => {
+                            return Err(Error::Custom(format!(
+                                "unexpected resolved type for atomic pointer operand: {other:?}"
+                            )))
+                        }
+                    }
+                };
```
```suggestion
                let pointer_space = func_ctx.info[pointer].ty
                    .inner_with(&module.types)
                    .pointer_space();
```

We have a few helper functions that we could use here.

ErichDonGubler replied:
👀 Onice, implemented more-or-less as suggested with 57a9728b.

teoxoy reviewed (commented):
The new test file looks much more complete than what we had before!

https://github.com/gfx-rs/naga/blob/da8e911d9d207a7571241e875f7ac066856e8b3f/tests/in/access.wgsl#L154-L169

I think we could remove this but the `atomicExchange` function seems to be missing from the new test file.

ErichDonGubler commented:
@teoxoy:

> The new test file looks much more complete than what we had before!

😊

> I think we could remove this [([`access.wgsl:154-169`](https://github.com/gfx-rs/naga/blob/da8e911d9d207a7571241e875f7ac066856e8b3f/tests/in/access.wgsl#L154-L169))] but the `atomicExchange` function seems to be missing from the new test file.

Agreed. Added coverage for `atomicExchange` with 340a6982 and b14bd408, and removed the overlapping contents of `access.wgsl` with f8c7cb12 (and threw in e3101219 for fun). Will autosquash momentarily so this branch's contents are still nice commits.

jimblandy commented on tests/out/hlsl/atomicOps.yaml:
What is this file doing here?

jimblandy reviewed (approved):
Looks good. Some minor cleanups needed.

jimblandy commented on tests/out/hlsl/atomicOps.hlsl:
```diff
@@ -0,0 +1,112 @@
+        workgroup_struct = (Struct)0;
+    }
+    GroupMemoryBarrierWithGroupSync();
+    storage_atomic_scalar.Store(0, asuint(1u));
+    storage_atomic_arr.Store(4, asuint(1));
+    storage_struct.Store(0, asuint(1u));
+    storage_struct.Store(4+4, asuint(1));
+    workgroup_atomic_scalar = 1u;
+    workgroup_atomic_arr[1] = 1;
+    workgroup_struct.atomic_scalar = 1u;
+    workgroup_struct.atomic_arr[1] = 1;
```
Connor found the following note in the [Direct3D functional spec](https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#7.14%20Shader%20Memory%20Consistency%20Model):

> atomic operations on UAV memory is atomic both with respect to other atomic operations as well as loads and stores.

Elsewhere, it also says:

> If a load or store operation spans more than 32-bits, the individual 32-bit operations are atomic, but not the whole.

I think this means that there's no need for special operations here: the proper translation of WGSL `atomicLoad` is indeed an ordinary HLSL `Load`.
